numpy/_core/src/umath/loops_utils.h.src

```diff
@@ -20,35 +20,36 @@
  * indicating negative steps.
  */
 NPY_FINLINE npy_bool
-nomemoverlap(char *ip, npy_intp ip_size, char *op, npy_intp op_size)
+nomemoverlap(char *ip, npy_intp ip_step, char *op, npy_intp op_step, npy_intp len)
 {
+    const npy_intp ip_size = ip_step*len, op_size = op_step*len;
     char *ip_start, *ip_end, *op_start, *op_end;
-    if (ip_size < 0) {
-        ip_start = ip + ip_size;
-        ip_end = ip;
+    if (ip_step < 0) {
+        ip_end = ip - ip_step;
+        ip_start = ip_end + ip_size;
```

Comment:
Looks right, thanks.  A "take it or leave it" comment if you like:  I am wondring if using `ip_offset = ip_step * (len - 1)` (rather than `ip_size`, just to avoid "size") wouldn't make this clearer, because the two branches would look identical.

Then do the calculation with inclusive ranges (and maybe a comment to that effect).

Reply:
Good suggestion! Done